src/lib/cache.ts

```diff
@@ -0,0 +1,24 @@
+import NodeCache from 'node-cache';
+
+export const cacheAsync = <T>(
+    fn: () => Promise<T>,
+    ttl: number,
+): [() => void, () => Promise<T>] => {
+    const myCache = new NodeCache();
+    const key = 'res';
+
+    const retFn = async () => {
+        const got = myCache.get(key);
+        if (got) {
+            return got as T;
+        }
+
+        const res = await fn();
```

Comment:
I think we're probably okay with this, but if we receive multiple requests simultaneously then they'll all end up calling `fn()` (since we're waiting for the async operation to complete before caching anything). We _could_ solve this by caching the promise immediately, and return that (and we'd have to then remove the promise from the cache if the operation fails). That's a bit of a change though in terms of the calling code, and there are pros/cons of each.

Reply:
I think this is okay as the impact is only a burst of requests before it settles.

Kind of related, but I've been wondering about the concurrency model of AWS Lambda in relation to this. They document some of this here:

https://docs.aws.amazon.com/lambda/latest/dg/invocation-scaling.html

But it makes it seem like lambda containers are never invoked concurrently - they deal with requests sequentially. Which seems very inefficient but maybe I've misread their docs!

Reply:
(If it is sequential, then this code is definitely fine.)

Reply:
okay https://read.iopipe.com/5-things-to-know-about-aws-lambda-running-lambda-functions-concurrently-1ee7ef1bf0b7 seems to confirm this view:

```
Every concurrent invocation of a Lambda function uses its own container. If you run out of “warm” containers (containers already spun up to run that Lambda), each concurrent invocation results in a cold start, adversely affecting performance.
```

Reply:
Cool thanks for sharing the links, I'll have a read! Didn't we read something before about things defined outside of the handler being shared across instances? Or maybe I'm mis-remembering. Either way I agree this is fine.

Reply:
State can be persisted across container *invocations* both in memory and in the `tmp` directory they provide, but you get no guarantees about how long a container will be around for.

Reply:
It isn't possible to share state across containers though I'm pretty sure (other than with an external to Lambda solution like a shared Redis or something).

Reply:
Ah yep that's right :+1: